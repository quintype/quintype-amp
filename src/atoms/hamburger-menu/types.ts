import { ReactElement } from "react";
import { DefaultTheme, CSSObject } from "styled-components";

export interface HamburgerMenuTypes {
  align: "left" | "right";
  textDirection: "ltr" | "rtl";
  items: MenuItem[];
  children?: JSX.Element[] | JSX.Element | ReactElement | HTMLElement;
  style?: CustomStyles;
  theme?: DefaultTheme;
}

interface MenuItem {
  title: string;
  url: string;
}

interface CustomStyles {
  list?: (theme) => CSSObject;
  listItem?: (theme) => CSSObject;
  listItemAnchor?: (theme) => CSSObject;
}
