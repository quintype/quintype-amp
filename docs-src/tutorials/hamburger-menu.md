The AMP hamburger menu takes items from a menu group. Which menu group that it takes its items from is decided based on the following logic:

- You can have items from any menu group of your choice. Just pass the slug of the menu group as a featureConfig to amp lib from your FE app and add items to that menu group. Check `sidebarMenu` in `feature config` in [opts tutorial]{@tutorial opts} for details
- else, if you've created a menu group having slug `amp-sidebar-menu` and added items to it, it'll pick those items
- else, if you've created a menu group having slug `sidebar-menu` and added items to it, it'll pick those items
- else it'll pick items from the `default` menu group
- else the sidebar menu is not shown

### Subdomains

Assume you are publisher `www.example.com` and have configured a subdomain `world-news`

Which menu group items are shown inside the hamburger menu on the subdomain is decided based on the following logic:

- 1st priority is given to featureConfig, just like before. So pass the following opts:
  ```js
  const exampleOptsObj = {
    // main domain opts
    featureConfig: {
      sidebarMenu: {
        menuGroupSlug: "general-links"
      }
    },
    "world-news": {
      // opts for "world-news" subdomain
      featureConfig: {
        sidebarMenu: {
          menuGroupSlug: "world-links"
        }
      }
    }
  };
  ```
  your main domain will now show items from menu group having slug `general-links` and your subdomain `https://world-news.example.com` will show items from menu group having slug `world-links`. In this example, you don't _have_ to pass featureConfig for the main domain. Following will also work just fine.
  ```js
  const exampleOptsObj = {
    "world-news": {
      // opts for "world-news" subdomain
      featureConfig: {
        sidebarMenu: {
          menuGroupSlug: "world-links"
        }
      }
    }
  };
  ```
- else, if you've created a menu group having slug `amp-sidebar-menu-world-news` and added items to it, it'll pick those items
- else, if you've created a menu group having slug `sidebar-menu-world-news` and added items to it, it'll pick those items
- else it'll pick items from the `default` menu group
- else the sidebar menu is not shown
