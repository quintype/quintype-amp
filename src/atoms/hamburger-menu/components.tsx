import styled from "styled-components";
import { Close } from "../icons";

export const StyledList = styled.ul`
  list-style: none;
  margin: 0;
  padding: 0 15px;
  min-width: 220px;
  font: ${(props) => {
    const fontFamily = props.theme.font.family.primary;
    const fontWeight = props.theme.font.weight.bold;
    const fontSize = props.theme.font.size.s;
    return `${fontWeight} ${fontSize} ${fontFamily}`;
  }};
`;
export const StyledListItem = styled.li`
  margin: 15px 0;
  color: ${(props) => props.theme.color.secondaryColor};
`;
export const StyledCloseIcon = styled.div`
  cursor: pointer;
  width: 1rem;
  margin: 0 10px 0 auto;
`;
export const StyledAnchor = styled.a`
  text-decoration: none;
  color: ${(props) => props.theme.color.secondaryColor};
`;
export const CloseButton = () => (
  <StyledListItem>
    <StyledCloseIcon role="button" tabIndex={0} on="tap:sidebar.close">
      <Close />
    </StyledCloseIcon>
  </StyledListItem>
);
