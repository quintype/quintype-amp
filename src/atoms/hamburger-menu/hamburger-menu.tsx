import React, { Fragment } from "react";
import { Helmet } from "react-helmet";
import { HamburgerMenuTypes } from "./types";
import styled from "styled-components";
import { Close } from "../icons";

export const HamburgerMenu = ({ align, items, textDirection }: HamburgerMenuTypes) => {
  return (
    <Fragment>
      <Helmet>
        <script async={undefined} custom-element="amp-sidebar" src="https://cdn.ampproject.org/v0/amp-sidebar-0.1.js" />
      </Helmet>
      <amp-sidebar id="sidebar" layout="nodisplay" side={align}>
        <StyledList dir={textDirection}>
          <CloseButton />
          {items.map((item, i) => (
            <StyledListItem key={i}>
              <StyledAnchor href={item.url}>{item.title}</StyledAnchor>
            </StyledListItem>
          ))}
        </StyledList>
      </amp-sidebar>
    </Fragment>
  );
};

const StyledList = styled.ul`
  list-style: none;
  margin: 0;
  padding: 0 15px;
  min-width: 220px;
  font: ${(props) => {
    const fontFamily = props.theme.font.family.primary;
    const fontWeight = props.theme.font.weight.bold;
    const fontSize = props.theme.font.size.s;
    return `${fontWeight} ${fontSize} ${fontFamily}`;
  }};
`;
const StyledListItem = styled.li`
  margin: 15px 0;
  color: ${(props) => props.theme.color.secondaryColor};
`;
const StyledCloseIcon = styled.div`
  cursor: pointer;
  width: 1rem;
  margin: 0 10px 0 auto;
`;
export const StyledAnchor = styled.a`
  text-decoration: none;
  color: ${(props) => props.theme.color.secondaryColor};
`;
const CloseButton = () => (
  <StyledListItem>
    <StyledCloseIcon role="button" tabIndex={0} on="tap:sidebar.close">
      <Close />
    </StyledCloseIcon>
  </StyledListItem>
);
