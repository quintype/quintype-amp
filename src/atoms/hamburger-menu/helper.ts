import { MenuItemTypes } from "../../types/config";

export const createTree = (items) => {
  const output: MenuItemTypes[] = [];
  const idsArr: number[] = []; // stores ids of items that are added to output obj
  items.forEach((item) => {
    if (item["parent-id"] === null) {
      output.push(item);
      idsArr.push(item.id);
    }
  });
};

// const k = {
//   "tag-name": null,
//   "entity-properties": null,
//   "collection-id": null,
//   "entity-slug": null,
//   "item-id": null,
//   rank: 7337,
//   title: "Home",
//   "item-type": "link",
//   "section-slug": null,
//   "tag-slug": null,
//   id: 7342,
//   "parent-id": null,
//   url: "https://barandbench.com",
//   "entity-name": null,
//   "collection-slug": null,
//   "section-name": null,
//   data: {
//     color: "#FFFFFF",
//     link: "https://barandbench.com"
//   }
// };

// export const HamburgerMenu2 = ({ align, items, textDirection }: HamburgerMenuTypes) => {
//   return (
//     <Fragment>
//       <Helmet>
//         <script async={undefined} custom-element="amp-sidebar" src="https://cdn.ampproject.org/v0/amp-sidebar-0.1.js" />
//         <script
//           async={undefined}
//           custom-element="amp-nested-menu"
//           src="https://cdn.ampproject.org/v0/amp-nested-menu-0.1.js"
//         />
//       </Helmet>
//       <amp-sidebar id="sidebar" layout="nodisplay" side={align}>
//         <amp-nested-menu layout="fill">
//           <StyledList dir={textDirection}>
//             <CloseButton />
//             {getChildItems(items, null).map((item) => {
//               const childItems = getChildItems(items, item.id);
//               return childItems.length > 0 ? (
//                 <ListItemWithIcon key={item.id}>
//                   <StyledAnchor href={item.url}>{item.title}</StyledAnchor>
//                   <span amp-nested-submenu-open={true}>></span>
//                   <div amp-nested-submenu={true}>
//                     <StyledList>
//                       <ListItem></ListItem>
//                     </StyledList>
//                   </div>
//                 </ListItemWithIcon>
//               ) : (
//                 <ListItem key={item.id}>
//                   <StyledAnchor href={item.url}>{item.title}</StyledAnchor>
//                 </ListItem>
//               );
//             })}
//           </StyledList>
//         </amp-nested-menu>
//       </amp-sidebar>
//     </Fragment>
//   );
// };

// const ListItemWithIcon = styled.li`
//   margin: 15px 0;
//   color: ${(props) => props.theme.color.secondaryColor};
//   display: flex;
//   justify-content: space-between;
//   align-items: center;
// `;
// const getChildItems = (items, parentId) => items.filter((item) => item["parent-id"] === parentId);
