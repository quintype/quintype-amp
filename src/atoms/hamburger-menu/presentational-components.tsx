import styled, { css } from "styled-components";

export const StyledList = styled.ul`
  list-style: none;
  margin: 0;
  padding: 0 15px;
  min-width: 220px;
  font: ${(props) => {
    const fontFamily = props.theme.font.family.primary;
    const fontWeight = props.theme.font.weight.bold;
    const fontSize = props.theme.font.size.s;
    return `${fontWeight} ${fontSize} ${fontFamily}`;
  }};
`;
const baseListItemStyles = css`
  margin: 15px 0;
  color: ${(props) => props.theme.color.secondaryColor};
`;
export const StyledListItem = styled.li`
  ${baseListItemStyles}
`;
export const SubmenuWrapper = styled.li`
  ${baseListItemStyles}
  display: flex;
  justify-content: space-between;
  align-items: center;
`;
export const StyledCloseIcon = styled.div`
  cursor: pointer;
  width: 1rem;
  margin: 0 0 0 auto;
`;
export const StyledAnchor = styled.a`
  text-decoration: none;
  color: ${(props) => props.theme.color.secondaryColor};
`;
