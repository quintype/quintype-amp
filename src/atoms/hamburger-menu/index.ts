import { HamburgerMenu } from "./hamburger-menu";
export { HamburgerMenu };
