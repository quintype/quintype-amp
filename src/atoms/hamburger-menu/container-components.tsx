import React from "react";
import { MenuItemComponentTypes, TreeNodeComponentTypes, SubMenuTypes } from "./types";
import { StyledListItem, StyledAnchor, StyledCloseIcon, StyledList, SubmenuWrapper } from "./presentational-components";

export const TreeNode = ({ item }: TreeNodeComponentTypes) => {
  return item["child-items"] && item["child-items"].length ? (
    <SubmenuWrapper>
      <MenuItem item={item} />
      <SubmenuOpen />
      <SubMenu childItems={item["child-items"]} />
    </SubmenuWrapper>
  ) : (
    <StyledListItem>
      <MenuItem item={item} />
    </StyledListItem>
  );
};

export const CloseButton = () => (
  <StyledListItem>
    <StyledCloseIcon role="button" tabIndex={0} on="tap:sidebar.close" aria-label="close">
      X
    </StyledCloseIcon>
  </StyledListItem>
);

const SubmenuOpen = () => <span amp-nested-submenu-open="">{" > "}</span>;
const SubmenuClose = () => <span amp-nested-submenu-close="">{" < "}</span>;

export const MenuItem = ({ item }: MenuItemComponentTypes) =>
  item["item-type"] === "placeholder" ? <Placeholder item={item} /> : <DefaultItem item={item} />;

export const Placeholder = ({ item }: MenuItemComponentTypes) => <span>{item.title}</span>;
export const DefaultItem = ({ item }: MenuItemComponentTypes) => (
  <StyledAnchor href={item.url || item.data.link}>{item.title}</StyledAnchor>
);

const SubMenu = ({ childItems }: SubMenuTypes) => (
  <div amp-nested-submenu="">
    <StyledList>
      {childItems.map((childItem) => (
        <TreeNode key={childItem.id} item={childItem} />
      ))}
      <SubmenuClose />
    </StyledList>
  </div>
);
