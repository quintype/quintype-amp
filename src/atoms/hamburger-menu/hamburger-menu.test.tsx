import React from "react";
import { HamburgerMenu } from "./hamburger-menu";
import { shallow } from "enzyme";

describe("HamburgerMenu", () => {
  it("should render", () => {
    const wrapper = shallow(<HamburgerMenu textDirection="ltr" align="left" items={[]} />);
    expect(wrapper).toMatchSnapshot();
  });
});
