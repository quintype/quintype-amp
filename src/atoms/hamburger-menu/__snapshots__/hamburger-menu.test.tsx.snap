// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HamburgerMenu should render 1`] = `
<Fragment>
  <HelmetWrapper
    defer={true}
    encodeSpecialCharacters={true}
  >
    <script
      custom-element="amp-sidebar"
      src="https://cdn.ampproject.org/v0/amp-sidebar-0.1.js"
    />
  </HelmetWrapper>
  <amp-sidebar
    id="sidebar"
    layout="nodisplay"
    side="left"
  >
    <styled.ul
      dir="ltr"
    />
  </amp-sidebar>
</Fragment>
`;
